### PR TITLE
drop async from benchmark

### PR DESCRIPTION
**Pull request by mrickard**
## Proposed Release Notes

Removed deprecated and failing suites from benchmark tests. 

## Links

https://issues.newrelic.com/browse/NR-40029

## Details

This doesn't yet remove async, so it's a draft PR. But it might be useful if you're looking to run the benchmark tests.

### CODE REVIEW

File: test/benchmark/shimmer/instrumentation.bench.js [hunk 1]
```diff
@@ -40,12 +40,13 @@ suite.add({
...
-  }
-})
+// This causes the cjs loader to fail, with benchmark not being found.
+// suite.add({
+//   name: 'shimmer.reinstrument()',
+//   agent: true,
+//   fn: function (agent) {
+//     return shimmer.reinstrument(agent, 'benchmark')
+//   }
+// })
```
Comment (mrickard) [inline test/benchmark/shimmer/instrumentation.bench.js:50]: I was inclined to remove this block, but it may be possible to have this run with the right path, if we really wanted to run it for some reason.

Comment (mrickard) [reply test/benchmark/shimmer/instrumentation.bench.js:50]: @jordigh The failure here is that `benchmark` can't be found from the context of the test, which is...weird. It's on my docket, along with the max-listeners-exceeded problem.

File: bin/run-bench.js [hunk 2]
```diff
@@ -5,7 +5,7 @@
 
 'use strict'
 
-const a = require('async')
+// const a = require('async')
```
[the remark this answers is not in the document]
Comment (mrickard) [reply bin/run-bench.js:8]: Yep. WIP; I'll rebase before merge.

File: bin/run-bench.js [hunk 3]
```diff
@@ -17,9 +17,17 @@ const tests = []
 const globs = []
 const opts = Object.create(null)
 
+const fakeCb = (err, payload) => {
+  if (err) {
```
[the remark this answers is not in the document]
Comment (mrickard) [reply bin/run-bench.js:21]: This is a kind of shim to keep the callback-heavy test style working as I convert over from `async`. Ultimately it'll be removed.

File: bin/run-bench.js [hunk 4]
```diff
@@ -74,67 +82,77 @@ class JSONPrinter {
...
+  const spawnEachFile = (file, spawnCb) => {
+    const test = path.relative(benchpath, file)
+
+    const args = [file]
+    if (opts.inspect) {
+      args.unshift('--inspect-brk')
+    }
+
+    // / TODO: remove diagnostic --trace-warnings
+    args.push('--trace-warnings')
```
[the remark this answers is not in the document]
Comment (mrickard) [reply bin/run-bench.js:126]: Yes--I wanted to see if I could find the source of the max-listeners-exceeded error, in case this approach is what triggered it. It turns out that the benchmark tests exceed the max listeners without my refactoring.

File: test/lib/benchmark.js [hunk 5]
```diff
@@ -35,104 +34,128 @@ class Benchmark {
...
     return (this.processedSamples = Object.keys(samples).reduce((acc, sampleName) => {
-      acc[sampleName] = new BenchmarkStats(samples[sampleName])
-      return acc
+      try {
+        acc[sampleName] = new BenchmarkStats(samples[sampleName])
+        return acc
+      } catch (e) {
+        // console.log(`********** ${sampleName} from ${this.name}`)
+        // console.log('this tests:', this.tests)
+        // console.log('err:', e)
```
Comment (mrickard) [inline test/lib/benchmark.js:43]: I'll remove these commented lines once I work out the db test errors.

File: bin/compare-bench-results.js [hunk 6]
```diff
@@ -18,119 +17,126 @@ if (process.argv.length !== 4) {
...
-      console.log(err)
-      process.exit(2)
+const processFile = async (file) => {
+  try {
+    const data = await fs.readFile(file, { encoding: 'utf8' })
+    let parsed = null
+    try {
+      parsed = JSON.parse(data)
+    } catch (parseError) {
+      return errorAndExit(parseError)
```
Comment (jmartin4563) [inline bin/compare-bench-results.js:33]: Since both these catches just `errorAndExit`, I don't think we need nested try's, we can just do one try/catch?

File: bin/run-bench.js [hunk 7]
```diff
@@ -19,9 +18,18 @@ const tests = []
...
+    console.error(err)
+    return
+  }
+  return payload
+}
+
 process.argv.slice(2).forEach(function forEachFileArg(file) {
   if (/^--/.test(file)) {
-    opts[file.substr(2)] = true
+    opts[file.substring(2)] = true
```
Comment (jmartin4563) [inline bin/run-bench.js:32]: Curious why this changed, they're semantically doing the same thing?

File: bin/run-bench.js [hunk 8]
```diff
@@ -76,67 +84,74 @@ class JSONPrinter {
 
 run()
 
-function run() {
+const errorAndExit = (err, message, code) => {
```
Comment (jmartin4563) [inline bin/run-bench.js:87]: We have two very similar implementations of this function, one here and one in compare-bench-results.js, can we just create one?

File: test/benchmark/shim/utilities.bench.js [hunk 9]
```diff
@@ -41,13 +41,6 @@ suite.add({
   }
 })
 
-suite.add({
```
Comment (jmartin4563) [inline test/benchmark/shim/utilities.bench.js:44]: Why was this removed?

File: test/benchmark/shimmer/instrumentation.bench.js [hunk 10]
```diff
@@ -40,12 +40,13 @@ suite.add({
...
 })
 
-suite.add({
-  name: 'shimmer.reinstrument()',
-  agent: true,
-  fn: function (agent) {
-    return shimmer.reinstrument(agent, 'benchmark')
-  }
-})
+// This causes the cjs loader to fail, with benchmark not being found.
```
Comment (jmartin4563) [inline test/benchmark/shimmer/instrumentation.bench.js:43]: Sorry to rehash a previous conversation, but did we log a ticket for fixing this?

File: test/lib/benchmark.js [hunk 11]
```diff
@@ -6,9 +6,8 @@
 'use strict'
 
 const helper = require('./agent_helper')
-const async = require('async')
 
-exports.createBenchmark = function createBenchmark(opts) {
+exports.createBenchmark = (opts) => {
```
Comment (jmartin4563) [inline test/lib/benchmark.js:10]: was this change actually needed?

**Review by jmartin4563 — CHANGES_REQUESTED** [review]
Couple of things, but `npm run bench` works for me now!

File: bin/run-bench.js (hunk 7 above)
Comment (mrickard) [reply bin/run-bench.js:32]: `substr` is a [deprecated method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr), so I went with the alternative that's likely to have a longer support timeline.

File: bin/run-bench.js (hunk 8 above)
Comment (mrickard) [reply bin/run-bench.js:87]: I was definitely leaning that way, but was biased toward not altering the expected behavior of the functions until I'd got them converted. Easy enough to do now!

File: test/benchmark/shim/utilities.bench.js (hunk 9 above)
Comment (mrickard) [reply test/benchmark/shim/utilities.bench.js:44]: @bizob2828 removed this method from the agent's shim interface with [this commit](https://github.com/newrelic/node-newrelic/commit/c470ff4fd223ba33a9c067463746b83918edd60d) in 2021; I documented that in the ticket comments. `events.EventEmitter.listenerCount` has apparently been deprecated from Node. Since this was removed from the agent code, we shouldn't be testing on it.

File: test/lib/benchmark.js (hunk 11 above)
Comment (mrickard) [reply test/lib/benchmark.js:10]: This one...not in terms of functionality. I've been converting function definitions to arrow functions where I can, so this was stylistically in keeping with that.

**Review by jmartin4563 — APPROVED** [review]
